### PR TITLE
add shared transaction constants

### PR DESCRIPTION
**Pull request by brad-decker**
### Description
Unifies and codifies that disparate constants used in various places through the app. We had some consts exported from the background that we used in the UI and some in the UI imported by the background. 

this adds a 'shared' folder at the root that can hold within it any utilities, constants, etc that we want to share between the two code bases.

### CODE REVIEW

File: app/scripts/controllers/transactions/tx-state-manager.js [hunk 1]
```diff
@@ -19,6 +19,7 @@ import { getFinalStates, normalizeTxParams } from './lib/util'
   <br>   - `'submitted'` the tx is sent to a server
   <br>   - `'confirmed'` the tx has been included in a block.
   <br>   - `'failed'` the tx failed for some reason, included on tx data.
+  <br>   - `'network-rejected'` the tx was rejected by the network for some reason.
```
Comment (brad-decker) [inline app/scripts/controllers/transactions/tx-state-manager.js:22]: This is an unrelated change, will be removed

File: app/scripts/controllers/transactions/index.js [hunk 2]
```diff
@@ -468,7 +455,7 @@ export default class TransactionController extends EventEmitter {
     } catch (err) {
       // this is try-catch wrapped so that we can guarantee that the nonceLock is released
       try {
-        this.txStateManager.setTxStatusFailed(txId, err)
+        this.txStateManager.setTxStatusNetworkRejected(txId, err)
```
Comment (brad-decker) [inline app/scripts/controllers/transactions/index.js:458]: This is an unrelated change and will be removed

File: shared/constants/transaction.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+// Transaction Group Category is a MetaMask construct to categorize the intent
+// of a group of transactions for purposes of displaying in the UI
+// 1. `'send'`              - Transaction group representing ether being sent from the user
+// 2. `'receive'`           - Transaction group representing a deposit/incoming transaction
+//                            usually maps 1:1 with TRANSACTION_CATEGORY_INCOMING
+// 3. `'interaction'`       - Transaction group representing an interaction with a smart
+//                            contract's methods.
+// 4. `'approval'`          - Transaction group representing a request for an allowance
+//                            of a token to spend on the user's behalf
+// 5. `'signature-request'` - Transaction group representing a signature request
```
Comment (Gudahtt) [inline shared/constants/transaction.js:72]: In an ideal world, these would not be included here. I don't think we should call anything a "transaction" that is not a transaction, in the code or the UI. It's unfortunate that this is how the signature requests were integrated. This is responsible for a lot of existing complicated UI logic.

It would seem more appropriate to have a higher-level abstraction (e.g. "activity"), something that could be either a transaction or a signature request/signing, or something else. The set of things we want to show in the activity tab.

File: shared/constants/transaction.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+//                    3. TRANSACTION_STATUS_SUBMITTED
+export const TRANSACTION_GROUP_STATUS_CANCELLED = 'cancelled'
+export const TRANSACTION_GROUP_STATUS_PENDING = 'pending'
+
+// Transaction Group Category is a MetaMask construct to categorize the intent
+// of a group of transactions for purposes of displaying in the UI
+// 1. `'send'`              - Transaction group representing ether being sent from the user
+// 2. `'receive'`           - Transaction group representing a deposit/incoming transaction
+//                            usually maps 1:1 with TRANSACTION_CATEGORY_INCOMING
+// 3. `'interaction'`       - Transaction group representing an interaction with a smart
```
Comment (Gudahtt) [inline shared/constants/transaction.js:68]: It might be worth clarifying that these are just the interactions that we haven't special-cased. `approve` is a contract interaction as well.

File: shared/constants/transaction.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+//                    2. TRANSACTION_STATUS_UNAPPROVED
+//                    3. TRANSACTION_STATUS_SUBMITTED
+export const TRANSACTION_GROUP_STATUS_CANCELLED = 'cancelled'
+export const TRANSACTION_GROUP_STATUS_PENDING = 'pending'
+
+// Transaction Group Category is a MetaMask construct to categorize the intent
+// of a group of transactions for purposes of displaying in the UI
+// 1. `'send'`              - Transaction group representing ether being sent from the user
+// 2. `'receive'`           - Transaction group representing a deposit/incoming transaction
+//                            usually maps 1:1 with TRANSACTION_CATEGORY_INCOMING
```
Comment (Gudahtt) [inline shared/constants/transaction.js:67]: "usually"? :thinking: I'd be curious to know when it doesn't line up

File: shared/constants/transaction.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+// Transaction Group Status is a MetaMask construct to track status of groups
+// of transactions.
+// 1. `'cancelled'` - A cancel type transaction in the group was confirmed
+// 2. `'pending'`   - The primaryTransaction of the group has a status that is one of:
+//                    1. TRANSACTION_STATUS_APPROVED
+//                    2. TRANSACTION_STATUS_UNAPPROVED
+//                    3. TRANSACTION_STATUS_SUBMITTED
+export const TRANSACTION_GROUP_STATUS_CANCELLED = 'cancelled'
+export const TRANSACTION_GROUP_STATUS_PENDING = 'pending'
```
Comment (Gudahtt) [inline shared/constants/transaction.js:61]: Should "confirmed" and "failed" be represented here as well?

File: shared/constants/transaction.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+export const TRANSACTION_TYPE_RETRY = 'retry'
+
+// Transaction Status is a mix of Ethereum terminology and MetaMask construct
+// 1. `'unapproved'` - A new transaction that the user has not approved or rejected
+// 2. `'approved'`   - The user has approved the transaction in the MetaMask UI
+// 3. `'rejected'`   - The user has rejected the transaction in the MetaMask UI
+// 4. `'signed'`     - The transaction has been signed
+// 5. `'submitted'`  - The transaction has been submitted
+// 6. `'failed'`     - The transaction has failed for some reason
+// 7. `'dropped'`    - The transaction was dropped due to a tx with same nonce being accepted
```
Comment (Gudahtt) [inline shared/constants/transaction.js:23]: Is that the only reason a transaction would be dropped? :thinking: I thought the network could drop it for other reasons as well.

File: shared/constants/transaction.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+export const TRANSACTION_STATUS_UNAPPROVED = 'unapproved'
+export const TRANSACTION_STATUS_REJECTED = 'rejected'
+export const TRANSACTION_STATUS_SIGNED = 'signed'
+export const TRANSACTION_STATUS_SUBMITTED = 'submitted'
+export const TRANSACTION_STATUS_FAILED = 'failed'
+export const TRANSACTION_STATUS_DROPPED = 'dropped'
+
+// Transaction Category is a MetaMask construct used internally
+// Used to categorize transactions by purpose for displaying information to the user
+// 1. `'tranfser'`            - A token transaction where the user is sending tokens that they
```
Comment (Gudahtt) [inline shared/constants/transaction.js:35]: Typo:
```suggestion
// 1. `'transfer'`            - A token transaction where the user is sending tokens that they
```

File: shared/constants/transaction.js (hunk 7 above)
Comment (brad-decker) [reply shared/constants/transaction.js:23]: Yeah lots of other reasons. I'm going to update this (and updating my own knowledge)

File: shared/constants/transaction.js (hunk 3 above)
Comment (brad-decker) [reply shared/constants/transaction.js:72]: Agreed. Although at this point in the code the things that get returned from nonceSortedTransactionSelector are called `TransactionGroup`, I intend on changing that in a future PR and updating these to reflect that. Also ideally I'll at some point move this logic into a background controller, enabling things like *real* pagination of activity, and that would make these variables -- whatever they end up being called -- make sense in a shared constant file.

File: shared/constants/transaction.js (hunk 6 above)
Comment (brad-decker) [reply shared/constants/transaction.js:63]: I *could* but these are the statuses that don't exist as transaction statuses. Honestly i could go either way with it. If we had separate consts for those it would enable us to change the string value for one without changing the other in the future. These two cases here represent areas where we override the status by inferring from the transaction data. For example, pending happens when the transaction is in one of the mentioned statuses, and cancelled happens when a cancel type transaction has been confirmed... the others are just inherited from the primary transaction.

File: shared/constants/transaction.js (hunk 8 above)
Comment (Gudahtt) [reply shared/constants/transaction.js:35]: This typo appears to still be here :thinking:

**Comment by brad-decker** [comment]
@Gudahtt I think this is ready for another review

**Comment by brad-decker** [comment]
@darkwing would you mind reviewing this, specifically that transactions selector file that you recently modified. I only updated the name of constants but it'd be good to get a thumbs up on it.

File: shared/constants/transaction.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
+// Transaction Type is a MetaMask construct used internally
+// There are three types at present
```
Comment (rekmarks) [inline shared/constants/transaction.js:2]: nit: redundant
```suggestion
```

File: shared/constants/transaction.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+export const TRANSACTION_STATUS_CONFIRMED = 'confirmed'
+export const TRANSACTION_STATUS_UNAPPROVED = 'unapproved'
+export const TRANSACTION_STATUS_REJECTED = 'rejected'
+export const TRANSACTION_STATUS_SIGNED = 'signed'
+export const TRANSACTION_STATUS_SUBMITTED = 'submitted'
+export const TRANSACTION_STATUS_FAILED = 'failed'
+export const TRANSACTION_STATUS_DROPPED = 'dropped'
+
+// Transaction Category is a MetaMask construct used internally
+// Used to categorize transactions by purpose for displaying information to the user
```
Comment (rekmarks) [inline shared/constants/transaction.js:34]:
```suggestion
// Transaction Category is a MetaMask construct used internally to
// categorize transactions by purpose for displaying information to the user.
```

File: shared/constants/transaction.js [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+export const TRANSACTION_CATEGORY_TOKEN_METHOD_APPROVE = 'approve'
+export const TRANSACTION_CATEGORY_TOKEN_METHOD_TRANSFER_FROM = 'transferfrom'
+export const TRANSACTION_CATEGORY_INCOMING = 'incoming'
+export const TRANSACTION_CATEGORY_SENT_ETHER = 'sentEther'
+export const TRANSACTION_CATEGORY_CONTRACT_INTERACTION = 'contractInteraction'
+export const TRANSACTION_CATEGORY_DEPLOY_CONTRACT = 'contractDeployment'
+export const TRANSACTION_CATEGORY_SWAP = 'swap'
+export const TRANSACTION_CATEGORY_SWAP_APPROVAL = 'swapApproval'
+
+// Transaction Group Status is a MetaMask construct to track status of groups
```
Comment (rekmarks) [inline shared/constants/transaction.js:62]: I am the nit king, I can do anything:
```suggestion
// Transaction Group Status is a MetaMask construct to track the status of groups
```

File: shared/constants/transaction.js (hunk 4 above)
Comment (rekmarks) [reply shared/constants/transaction.js:77]: Was this comment addressed?

File: ui/app/hooks/useSwappedTokenValue.js [hunk 12]
```diff
@@ -12,7 +12,7 @@ import { useTokenFiatAmount } from './useTokenFiatAmount'
  */
 
 /**
- * A SWAP transaction group's primaryTransaction contains details of the swap,
+ * A Swap transaction group's primaryTransaction contains details of the swap,
```
Comment (rekmarks) [inline ui/app/hooks/useSwappedTokenValue.js:15]: 👏

**Review by rekmarks — CHANGES_REQUESTED** [review]
I have some nits about documentation that I think are worthwhile, but otherwise I am very happy about this PR.

My one general piece of feedback is that I think it's more ergonomic to define objects of constants and then export that object, rather than adding multiple exports.

For example that'd allow us to do something like:
```js
import { TRANSACTION_CATEGORIES } from '../shared/constants/transaction'
// ...
if (tx.category === TRANSACTION_CATEGORIES.SWAP) { /* ... */ }
```

Also, if you ever need a map of constant values, you already have it.

This is just a suggestion, and I won't block on it!

**Review by rekmarks — DISMISSED** [review]
Very nice!

**Comment by brad-decker** [comment]
@rekmarks thanks for the suggestion to export objects with keys instead of individual variables. It worked out so well and now, by using `@typedef` for the object declaration, we have very nice developer tooling in vscode:

image

**Review by rekmarks — APPROVED** [review]
Great work with the typings. LGTM! 🙌

File: ui/app/helpers/utils/transactions.util.js [hunk 13]
```diff
@@ -184,8 +180,8 @@ export function getStatusKey(transaction) {
   }
 
   if (
-    status === TRANSACTION_STATUS_CONFIRMED &&
-    type === TRANSACTION_TYPE_CANCEL
+    status === TRANSACTION_STATUSES.CONFIRMED &&
+    type === TRANSACTION_TYPES.CANCEL
```
Comment (Gudahtt) [inline ui/app/helpers/utils/transactions.util.js:184]: The hard-coded `failed` and `cancelled` values returned here should be using the constants as well

Comment (Gudahtt) [reply ui/app/helpers/utils/transactions.util.js:184]: We should use these constants in `tx-state-manager` as well :thinking: That's where the `status` returned below is set.

**Review by Gudahtt — COMMENTED** [review]
Belated, but, LGTM!

I noticed a few things that can be addressed in a future PR (a typo, and a few places where these constants should be used instead of string literals)

**Comment by brad-decker** [comment]
@Gudahtt i'll fix these today! Thanks for pointing them out
